### PR TITLE
Panorama preview weapons

### PR DESCRIPTION
**Pull request by Nibuja05**
Added panorama preview weapons: They are similar to normal panorama weapons, but cannot be dragged and don't need a specific weaponID. Instead they receive their data directly. This allows us, to display weapons in panorama before creating them (only from their data).
Additionally:
- added team assignment
- fixed broken hero_spin (for multiple players)
- added more panorama documentation

### CODE REVIEW

File: content/panorama/scripts/custom_game/util.ts [hunk 1]
```diff
@@ -13,3 +18,78 @@ function print(...args: any[]) {
...
+	return "<font color='#ffffff'><b>" + text + "</b></font>";
+}
+
+/**
+ * Removes the highlight of a text part through html formatting.
+ * @param text text to remove the highlight
+ * @returns formatted text
+ */
+function UnmarkText(text: string | number): string {
+	return "<font color='#4b535e'><b>" + text + "</b></font>";
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/util.ts:37]: If you use MarkText and then UnmarkText one after the other, wouldn't that text have two layers of HTML and <b>? That sounds like something that could go wrong.....
I would suggest making it also remove a marked text (using substrings maybe?) first before adding the standard unmarked HTML.

File: content/panorama/scripts/custom_game/util.ts [hunk 2]
```diff
@@ -13,3 +18,78 @@ function print(...args: any[]) {
...
+}
+
+/**
+ * Shortens a floting point number for better readbility.
+ * @param num number to shorten
+ * @returns string representing the shortened number
+ */
+function ShortenNumber(num: number) {
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/util.ts:45]: Wouldn't it be easier to just do Math.round() on the number? That sounds like a really weird way to shorten a number. You can also include an additional argument to decide how many decimal points (defaults to 2).

In either case, the number shouldn't be returned as a string - it makes no sense to return a number in string form.

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 3]
```diff
@@ -30,28 +44,55 @@ class PanoramaWeaponBase {
 		this.panel.weaponBaseRef = this;
 	}
 
-	protected isValidDropTarget(targetPanel: Panel, displayPanel: WeaponDragPanel): boolean {
+	/**
+	 * Can a weapon be dropped here?
+	 * @param displayPanel temporary dragged panel
+	 * @returns always true
+	 */
+	protected isValidDropTarget(displayPanel: WeaponDragPanel): boolean {
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/weapon_base.ts:52]: What's the point of that function if it's always returning true?

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 4]
```diff
@@ -138,11 +199,15 @@ class PanoramaWeaponSlot extends PanoramaWeaponBase {
...
+	/**
+	 * Set the type of this panel (defines size).
+	 * @param panelType type of this panel.
+	 */
 	public setPanelType(panelType: PanoramaWeaponPanelType) {
 		this.panelType = panelType;
 	}
 
-	protected isValidDropTarget(targetPanel: Panel, displayPanel: WeaponDragPanel): boolean {
+	protected isValidDropTarget(displayPanel?: WeaponDragPanel): boolean {
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/weapon_base.ts:210]: This function isn't utilizing the `displayPanel` argument, regardless of if it is defined or not. Is it needed here?

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 5]
```diff
@@ -192,12 +271,174 @@ class PanoramaWeapon {
...
+	 */
+	protected getValues(): CurrentWeaponValue[] {
+		return [];
+	}
+
+	/**
+	 * Retrieves all needes data to display a tooltip for a weapon and creates according panels.
+	 * @param displayPanel panel to show the tooltip on
+	 * @param name name of the weapon
+	 * @param id `weaponID` of the weapon
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/weapon_base.ts:314]: ID is not an argument to this function anymore.

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 6]
```diff
@@ -192,12 +271,174 @@ class PanoramaWeapon {
...
+
+		let tooltipWeaponValues = $.CreatePanel("Label", $.GetContextPanel(), "WeaponValues");
+		tooltipWeaponValues.AddClass("TooltipWeaponValues");
+		tooltipWeaponValues.SetParent(tooltipWeaponContent);
+		tooltipWeaponValues.html = true;
+		let finalText: string[] = [];
+		for (let value of Object.values(values)) {
+			let curVal = ShortenNumber(value.cur_val);
+			let nextVal = ShortenNumber(value.next_val);
+			let text =
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/weapon_base.ts:366]: You might want to use string literals (using backticks) for better readability here. Less `+` and `"` in general.

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 7]
```diff
@@ -321,133 +576,39 @@ class PanoramaWeapon {
...
+class PanoramaPreviewWeapon extends PanoramaWeaponBase {
+	element: WeaponElement;
+	values: CurrentWeaponValue[];
+
+	/**
+	 * @param name name of this weapon
+	 * @param level level of this weapon
+	 * @param icon icon of this weapon
+	 * @param color color of this weapon (for some panorama interactions)
+	 * @param id `weaponID` of this weapon (to refence values)
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/weapon_base.ts:588]: No ID here

File: content/panorama/scripts/custom_game/welcome.ts [hunk 8]
```diff
@@ -43,9 +48,17 @@ GameEvents.Subscribe("load_preset", (event) => {
 	let mainPanel = $("#WelcomeMain").GetParent()!;
 	mainPanel.AddClass("WelcomeVisible");
 	mainPanel.RemoveClass("WelcomeHidden");
-	let weapons: PanoramaWeapon[] = [];
-	for (let [_, value] of Object.entries(event.weapons)) {
-		let newWeapon = new PanoramaWeapon(value.name, value.level, value.icon, value.color, value.id);
+	let weapons: PanoramaPreviewWeapon[] = [];
+	for (let [_, weapon] of Object.entries(event.weapons)) {
+		// let newWeapon = new PanoramaWeapon(value.name, value.level, value.icon, value.color, value.id);
```
Comment (Shushishtok) [inline content/panorama/scripts/custom_game/welcome.ts:53]: If not needed, remove print

File: game/scripts/vscripts/GameMode.ts [hunk 9]
```diff
@@ -156,24 +157,38 @@ export class GameMode {
 				}
 			}
 		}
-		if (unit.IsRealHero() && unit.GetUnitName() == GenericUnitName.PlayerHero) {
+		// if (unit.IsRealHero() && unit.GetUnitName() == GenericUnitName.PlayerHero) {
```
Comment (Shushishtok) [inline game/scripts/vscripts/GameMode.ts:160]: If comment's not needed, remove

File: game/scripts/vscripts/core/game_mode_manager.ts [hunk 10]
```diff
@@ -93,56 +96,86 @@ function FirstPerk(hero: CDOTA_BaseNPC_Hero) {
 }
 
 function ShowPresets(playerID: PlayerID) {
-	print("Show Presets!");
+	const player = PlayerResource.GetPlayer(playerID);
+	if (!player) error(`Player with id [${playerID}] could not be found.`);
 	const chosenPresets = Presets.GetRandomPresetCountUnique(4);
+	let thesePresets: { [index: number]: { name: string; weaponData: WeaponData[] } } = {};
```
Comment (Shushishtok) [inline game/scripts/vscripts/core/game_mode_manager.ts:102]: Kind of a weird naming I think? Maybe a better name? (not `thosePresets` lol)

File: game/scripts/vscripts/core/game_mode_manager.ts [hunk 11]
```diff
@@ -93,56 +96,86 @@ function FirstPerk(hero: CDOTA_BaseNPC_Hero) {
...
+	currentPresets.set(playerID, thesePresets);
+}
+
+/**
+ * Generates level specific special values from raw weaponData. Same as `generateValues` from `BaseWeapon`.
+ * @param data raw weapon data
+ * @param weaponLevel weapon level
+ * @returns generated weapon values
+ */
+function GenerateValuesFromWeaponData(data: WeaponData, weaponLevel: number): CurrentWeaponValue[] {
```
Comment (Shushishtok) [inline game/scripts/vscripts/core/game_mode_manager.ts:148]: I think this belongs in the Weapon Manager namespace.

File: game/scripts/vscripts/core/hero_spin.ts [hunk 12]
```diff
@@ -60,14 +60,15 @@ export class spin_passive extends SpinModifier {
 	}
 
 	OnCreated(): void {
+		if (IsClient()) return;
+		if (!this.parent.IsRealHero()) {
```
Comment (Shushishtok) [inline game/scripts/vscripts/core/hero_spin.ts:64]: Is this actually needed, now that only real heroes get the ability?

File: game/scripts/vscripts/lib/util.ts [hunk 13]
```diff
@@ -467,3 +466,31 @@ function CanUnitCastAbility(caster: CDOTA_BaseNPC, ability: CDOTABaseAbility, ta
...
+		case 5:
+			return DotaTeam.CUSTOM_5;
+		case 6:
+			return DotaTeam.CUSTOM_6;
+		case 7:
+			return DotaTeam.CUSTOM_7;
+		case 8:
+			return DotaTeam.CUSTOM_8;
+		default:
+			return DotaTeam.NOTEAM;
```
Comment (Shushishtok) [inline game/scripts/vscripts/lib/util.ts:494]: This should never occur, and therefore should throw an error if this happens.

**Review by Shushishtok — CHANGES_REQUESTED** [review]

File: content/panorama/scripts/custom_game/util.ts (hunk 1 above)
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/util.ts:37]: Yes, this behavior is wanted like this, jsut the name might be confusing.
This function is used for creating unmarked text inside marked text; there wouldn't be any html to remove.
Im going to rename it, so it matches its purpose more

File: content/panorama/scripts/custom_game/weapon_base.ts (hunk 3 above)
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/weapon_base.ts:52]: it is the default return value.
It gets overridden in the child classes.

File: content/panorama/scripts/custom_game/weapon_base.ts (hunk 4 above)
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/weapon_base.ts:210]: yes, because its called like this: `if (this.isValidDropTarget(displayPanel)) {`

**Review by Shushishtok — APPROVED** [review]

**Pull request merged**